### PR TITLE
[PTX-1454] Added GKE implementation to set and get InstanceGroup Count

### PR DESCRIPTION
**Pull request by vinayakshnd**
1. Added SetInstanceGroupSize() and GetClusterSize() compute APIs and their corresponding 
    implementation for GKE.
2. Added test cases for above APIs

### CODE REVIEW

File: cloudops.go [hunk 1]
```diff
@@ -48,6 +52,14 @@ type Compute interface {
 	// InspectInstanceGroupForInstance inspects the instance group to which the
 	// cloud instance with given ID belongs
 	InspectInstanceGroupForInstance(instanceID string) (*InstanceGroupInfo, error)
+	// SetInstanceGroupSize sets desired node count per availability zone
+	// for given instance group
+	SetInstanceGroupSize(instanceGroupInfo *InstanceGroupInfo,
```
Comment (vinayakshnd) [inline cloudops.go]: Pass instanceGroupID and location of instance group. Figure out ClusterName internally.

File: cloudops.go [hunk 2]
```diff
@@ -48,6 +52,14 @@ type Compute interface {
...
+		count int64,
+		timeout time.Duration) error
+	// GetClusterSize returns current node count in given cluster
+	// This count is total node count accross all availability zones
+	GetClusterSize(instanceGroupInfo *InstanceGroupInfo) (int64, error)
```
Comment (vinayakshnd) [inline cloudops.go]: GetClusterSizeForInstance(instanceID) 
Add getClusterName() API

File: gce/gce.go [hunk 3]
```diff
@@ -213,16 +213,20 @@ func (s *gceOps) InspectInstanceGroupForInstance(instanceID string) (*cloudops.I
...
+			instanceGroupMetadata := make(map[string]interface{})
+			instanceGroupMetadata[clusterNameKey] = gkeClusterName
+
 			retval := &cloudops.InstanceGroupInfo{
 				CloudResourceInfo: cloudops.CloudResourceInfo{
 					Name:   nodePool.Name,
-					Zone:   s.inst.zone,
-					Region: s.inst.region,
+					Zone:   clusterLocation,
+					Region: clusterLocation,
```
Comment (vinayakshnd) [inline gce/gce.go]: Correctly populate this

File: test/cloudops.go [hunk 4]
```diff
@@ -59,6 +64,36 @@ func compute(t *testing.T, driver cloudops.Ops) {
 	groupInfo, err := driver.InspectInstanceGroupForInstance(instanceID)
 	require.NoError(t, err, "failed to inspect instance group")
 	require.NotNil(t, groupInfo, "got nil instance group info from inspect")
```
Comment (vinayakshnd) [inline test/cloudops.go]: test with both regional and zonal clusters.

File: cloudops.go [hunk 5]
```diff
@@ -32,6 +34,8 @@ type InstanceGroupInfo struct {
 	Max *int64
 	// Zones that the instance group is part of
 	Zones []string
+	// Metadata holds additional information regarding instance group
+	Metadata map[string]interface{}
```
Comment (vinayakshnd) [inline cloudops.go]: Remove this generic map. Get ClusterName explicitly.

File: test/cloudops.go (hunk 4 above)
Comment (vinayakshnd) [reply test/cloudops.go]: Tested these changes with both regional (spread across multiple zone) and zonal (single zone) clusters.

File: cloudops.go (hunk 1 above)
Comment (vinayakshnd) [reply cloudops.go]: Updated SetInstanceGroupSize() to only take instanceGroupID and location

File: gce/gce.go [hunk 6]
```diff
@@ -572,6 +572,102 @@ func (s *gceOps) RemoveTags(
...
+	}
+	cluster, err := s.containerService.Projects.Zones.Clusters.Get(s.inst.project, groupInfo.Zone, clusterName).Do()
+	if err != nil {
+		return int64(0), err
+	}
+
+	return cluster.CurrentNodeCount, nil
+}
+
+func (s *gceOps) GetClusterName() (string, error) {
```
Comment (harsh-px) [inline gce/gce.go]: this doesn't need to be public

File: gce/gce.go [hunk 7]
```diff
@@ -572,6 +572,102 @@ func (s *gceOps) RemoveTags(
...
+		if item == nil {
+			continue
+		}
+
+		if item.Key == clusterNameKey {
+			if item.Value == nil {
+				return "", fmt.Errorf("instance has %s key in metadata but has invalid value", clusterNameKey)
+			}
+
+			gkeClusterName = *item.Value
```
Comment (harsh-px) [inline gce/gce.go]: just return from here.

File: gce/gce.go [hunk 8]
```diff
@@ -572,6 +572,102 @@ func (s *gceOps) RemoveTags(
...
+		}
+	}
+	return gkeClusterName, nil
```
Comment (harsh-px) [inline gce/gce.go]: if you fall through here, return err that cluster name wasn't found.

File: gce/gce.go [hunk 9]
```diff
@@ -572,6 +572,102 @@ func (s *gceOps) RemoveTags(
...
+				logrus.Infof("cluster node count of [%s] matches with expected count [%v]. Exiting", clusterName, clusterInfo.CurrentNodeCount)
+				return nil, false, nil
+			}
+
+			return nil,
+				true,
+				fmt.Errorf("cluster node count of [%s] does not match. Expected: [%v], Actual:[%v]. Waiting", clusterName, count*int64(len(clusterInfo.Locations)), clusterInfo.CurrentNodeCount)
+		}
+
+		_, err = task.DoRetryWithTimeout(f, timeout, 30*time.Second)
```
Comment (harsh-px) [inline gce/gce.go]: Use a shorter interval. Change to 15 seconds. And define at const at top of file.

Comment (harsh-px) [inline gce/gce.go]: just `return task.DoRetryWithTimeout(f, timeout, 30*time.Second)` directly.

File: gce/gce.go [hunk 10]
```diff
@@ -572,6 +572,102 @@ func (s *gceOps) RemoveTags(
...
+			}
+
+			if clusterInfo.CurrentNodeCount == count*int64(len(clusterInfo.Locations)) {
+				logrus.Infof("cluster node count of [%s] matches with expected count [%v]. Exiting", clusterName, clusterInfo.CurrentNodeCount)
+				return nil, false, nil
+			}
+
+			return nil,
+				true,
+				fmt.Errorf("cluster node count of [%s] does not match. Expected: [%v], Actual:[%v]. Waiting", clusterName, count*int64(len(clusterInfo.Locations)), clusterInfo.CurrentNodeCount)
```
Comment (harsh-px) [inline gce/gce.go]: split `fmt.Errorf` into multiple lines to avoid wrapping.

Same for other lines in the review

File: gce/gce.go [hunk 11]
```diff
@@ -572,6 +572,102 @@ func (s *gceOps) RemoveTags(
...
+
+	if timeout > time.Nanosecond {
+		f := func() (interface{}, bool, error) {
+			clusterInfo, err := s.containerService.Projects.Zones.Clusters.Get(s.inst.project, instanceGroupLocation, clusterName).Do()
+			if err != nil {
+				return nil, true, err
+			}
+
+			if clusterInfo.CurrentNodeCount == count*int64(len(clusterInfo.Locations)) {
+				logrus.Infof("cluster node count of [%s] matches with expected count [%v]. Exiting", clusterName, clusterInfo.CurrentNodeCount)
```
Comment (harsh-px) [inline gce/gce.go]: Remove this log line. Don't need prints in the happy path.

File: test/cloudops.go [hunk 12]
```diff
@@ -59,6 +64,36 @@ func compute(t *testing.T, driver cloudops.Ops) {
...
+
+	err = driver.SetInstanceGroupSize(groupInfo.Name, groupInfo.Zone, clusterNodeCount, 5*time.Minute)
+	if err != nil {
+		_, ok := err.(*cloudops.ErrNotSupported)
+		if !ok {
+			t.Errorf("Fialed to set node count. Error:[%v]", err)
```
Comment (harsh-px) [inline test/cloudops.go]: this has a typo: Fialed

File: backoff/exponential.go [hunk 13]
```diff
@@ -89,6 +89,43 @@ func (e *exponentialBackoff) InspectInstanceGroupForInstance(instanceID string)
...
+}
+
+func (e *exponentialBackoff) GetClusterSizeForInstance(instanceID string) (int64, error) {
+	var (
+		count   int64
+		origErr error
+	)
+	conditionFn := func() (bool, error) {
+		count, origErr = e.cloudOps.GetClusterSizeForInstance(instanceID)
+		msg := fmt.Sprintf("Failed to get instance group size for: %v.", instanceID)
```
Comment (harsh-px) [inline backoff/exponential.go]: should get `get cluster size`

Comment (harsh-px) [inline backoff/exponential.go]: don't need to save msg in a variable.

File: gce/gce.go (hunk 9 above)
Comment (vinayakshnd) [reply gce/gce.go]: Hi @harsh-px ,

`task.DoRetryWithTimeout(f, timeout, 30*time.Second)` gives us interface{} and err.
Hence we can't directly return.

File: test/cloudops.go [hunk 14]
```diff
@@ -59,6 +64,36 @@ func compute(t *testing.T, driver cloudops.Ops) {
...
+	}
+
+	// Validate when timeout is given as 0, API does not error out.
+	err = driver.SetInstanceGroupSize(groupInfo.Name, groupInfo.Zone, clusterNodeCount+1, 0)
+	if err != nil {
+		_, ok := err.(*cloudops.ErrNotSupported)
+		if !ok {
+			t.Errorf("Fialed to set node count. Error:[%v]", err)
+		}
+	}
```
Comment (harsh-px) [inline test/cloudops.go]: the UT doesn't check if the instance group size actually got updated. You need to add a GetClusterSize.. call after that.

**Comment by vinayakshnd** [comment]
Hi @harsh-px I have also added support for regional clusters in this PR.

File: test/cloudops.go [hunk 15]
```diff
@@ -59,6 +70,74 @@ func compute(t *testing.T, driver cloudops.Ops) {
...
+
+	// Wait for count to get updated for an instance group
+	expectedNodeCount := (clusterNodeCount + 1) * int64(len(groupInfo.Zones))
+	f := func() (interface{}, bool, error) {
```
Comment (harsh-px) [inline test/cloudops.go]: skip this wait loop if the Set call above returns err not supported.

File: test/cloudops.go [hunk 16]
```diff
@@ -59,6 +70,74 @@ func compute(t *testing.T, driver cloudops.Ops) {
...
+		require.Equal(t, int64(clusterNodeCount*len(groupInfo.Zones)), currentCount,
+			"expected cluster node count does not match with actual node count")
+	}
+
+	// Validate when timeout is given as 0, API does not error out.
+	err = driver.SetInstanceGroupSize(groupInfo.Name, clusterLocation, clusterNodeCount+1, 0)
+	if err != nil {
+		_, ok := err.(*cloudops.ErrNotSupported)
+		if !ok {
+			t.Errorf("Fialed to set node count. Error:[%v]", err)
```
Comment (harsh-px) [inline test/cloudops.go]: typo Fialed here

**Pull request merged**